streamz/core.py

```diff
@@ -319,18 +339,35 @@ def remove_stream(change):
 
         return output._ipython_display_(**kwargs)
 
-    def _emit(self, x):
+    def _emit(self, x, metadata=None):
+        if metadata is None:
+            metadata = {}
+
+        refs = []
+        if 'refs' in metadata:
+            refs = metadata['refs']
+            if not isinstance(refs, list):
+                refs = [refs]
+
         result = []
-        for downstream in list(self.downstreams):
-            r = downstream.update(x, who=self)
+        downstreams = list(self.downstreams)
+        for ref in refs:
+            ref.retain(len(downstreams))
+
+        for downstream in downstreams:
+            r = downstream.update(x, who=self, metadata=metadata)
+
             if type(r) is list:
                 result.extend(r)
             else:
                 result.append(r)
 
+            for ref in refs:
+                ref.release()
```

Comment:
So even in completely sync operations, the references value goes up while the chain of tasks is executed, and then down again. This is unnecessary computing, but I suppose insignificant?

Reply:
I believe this is necessary because the source or creator of the ref counter cannot make assumption about what the pipeline looks like. The source knows that the data is done due to the value of the reference counter changing from greater than zero to zero. If we do not use counters for the sync operations, then the counter will never trigger the done callback.